ee/hogai/django_checkpoint/checkpointer.py

```diff
@@ -21,16 +21,32 @@
 from langgraph.checkpoint.serde.types import TASKS, ChannelProtocol
 
 from ee.models.assistant import ConversationCheckpoint, ConversationCheckpointBlob, ConversationCheckpointWrite
-from posthog.warehouse.util import database_sync_to_async
+from posthog.sync import database_sync_to_async
 
 
 class DjangoCheckpointer(BaseCheckpointSaver[str]):
     jsonplus_serde = JsonPlusSerializer()
-    _lock: asyncio.Lock
+    _thread_locks: dict[str, asyncio.Lock]
+    _locks_dict_lock: asyncio.Lock
 
     def __init__(self, *args):
         super().__init__(*args)
-        self._lock = asyncio.Lock()
+        self._thread_locks = {}
+        self._locks_dict_lock = asyncio.Lock()
+
+    async def _get_thread_lock(self, thread_id: str) -> asyncio.Lock:
+        """Get the async lock for a specific thread."""
+        async with self._locks_dict_lock:
+            if thread_id not in self._thread_locks:
+                self._thread_locks[thread_id] = asyncio.Lock()
+            return self._thread_locks[thread_id]
+
+    async def _cleanup_thread_lock(self, thread_id: str):
+        """Clean up thread lock if not currently locked."""
+        async with self._locks_dict_lock:
+            lock = self._thread_locks.get(thread_id)
+            if lock is not None and not lock.locked():
+                self._thread_locks.pop(thread_id, None)
```

Comment:
Nit: this would make for an easier API if we combined the two function into one, which would return an async `contextmanager` handling the `finally:` bit internally (as currently `_get_thread_lock()` consumers need to remember to implement their own `finally:`)